### PR TITLE
[RLlib] Fixing the missing LayerNorm bug in RLModuleSpec

### LINKED ISSUE #60926: [RLlib] `hidden_layer_use_layernorm` doesn't get applied when set.
### What happened + What you expected to happen

The `hidden_layer_use_layernorm` attribute should apply a LayerNorm layer to the model architecture when set, but it does not. This is because catalog.py fails to apply the setting when instantiating an `MLPEncoderConfig` on [line 292](https://github.com/ray-project/ray/blob/master/rllib/core/models/catalog.py#L292). The fix I'm presently using is below:

```
...
encoder_config = MLPEncoderConfig(
                    input_dims=observation_space.shape,
                    hidden_layer_dims=hidden_layer_dims,
                    hidden_layer_activation=activation,
                    hidden_layer_use_layernorm=model_config_dict[
                        "fcnet_use_layernorm"
                    ],
                    ...
```

I should note that `fcnet_use_layernorm` should also be added to [default_model_config.py](https://github.com/ray-project/ray/blob/4ab83f0d598cd44e43c9eb2225ee49efa227d231/rllib/core/rl_module/default_model_config.py).

There's an additional, related issue in line 163 of primitives.py that causes encoders with layer normalization not to use it on their last layer _(since it uses the same logic as an action or value head, which shouldn't do so)_:

https://github.com/ray-project/ray/blob/master/rllib/core/models/torch/primitives.py#L163

I would solve this by applying layer normalization so long as an output layer activation is used, since it seems like that's always what you'd want to do:

```
class TorchMLP(nn.Module):
    def __init__(... ):
        super().__init__()
        assert input_dim > 0

        self.input_dim = input_dim

        hidden_activation = get_activation_fn(
            hidden_layer_activation, framework="torch"
        )
        output_activation = get_activation_fn( # Moved this up
            output_activation, framework="torch"
        )
       ...

            # Changed this section
            activation = output_activation if is_output_layer else hidden_activation
            # Insert a layer normalization in between layer's output and
            # the activation.
            if hidden_layer_use_layernorm and (activation is not None):
                # We use an epsilon of 0.001 here to mimick the Tf default behavior.
                layers.append(nn.LayerNorm(dims[i + 1], eps=0.001))
            # Add the activation function.
            if activation is not None:
                layers.append(activation())
                
        self.mlp = nn.Sequential(*layers)
...
```

I'm happy to submit a PR if this fix looks right to you.

### Versions / Dependencies

Ray 2.53.0
Python 3.12.12
Ubuntu 22.04.5 LTS

### Reproduction script

```
from ray.rllib.algorithms.ppo.torch.default_ppo_torch_rl_module import (  # noqa
    DefaultPPOTorchRLModule,
)
from ray.rllib.core.rl_module.rl_module import RLModuleSpec
import gymnasium as gym

RLModuleSpec(
    module_class=DefaultPPOTorchRLModule,
    observation_space=gym.spaces.Box(0,1,(10,)),
    action_space=gym.spaces.Box(0,1,(10,)),
    model_config={
        #
        "fcnet_hiddens": [16] * (3-1),
        "head_fcnet_hiddens": [16] * (3-1),
        "vf_share_layers": False, 
        "fcnet_use_layernorm": True,
        "fcnet_activation": 'relu',
    },
).build()

# no LayerNorm layers will be added without the patches above
```

### PR DESCRIPTION
**Pull request by HassamSheikh**
Fixes #60926 

This PR fixes the issue of absence of LayerNorm layer in the RLModuleSpec as identified by OSS contributor

### CODE REVIEW

**Review by pseudo-rnd-thoughts — APPROVED** [review]
LGTM. @HassamSheikh could you comment on Simon's PR about unifying the catalog to ensure this isn't accidentally removed

**Comment by MatthewCWeston** [comment]
@HassamSheikh Thoughts on the fix for layer normalization being missing in the last layer of the encoder? From the original issue (#60926 ):

```
class TorchMLP(nn.Module):
    def __init__(... ):
        super().__init__()
        assert input_dim > 0

        self.input_dim = input_dim

        hidden_activation = get_activation_fn(
            hidden_layer_activation, framework="torch"
        )
        output_activation = get_activation_fn( # Moved this up
            output_activation, framework="torch"
        )
       ...

            # Changed this section
            activation = output_activation if is_output_layer else hidden_activation
            # Insert a layer normalization in between layer's output and
            # the activation.
            if hidden_layer_use_layernorm and (activation is not None):
                # We use an epsilon of 0.001 here to mimick the Tf default behavior.
                layers.append(nn.LayerNorm(dims[i + 1], eps=0.001))
            # Add the activation function.
            if activation is not None:
                layers.append(activation())
                
        self.mlp = nn.Sequential(*layers)
...
```

Without this change, an encoder won't apply layer normalization to the layer before the action/value heads. E.g. an actor with layer normalization will look like this:

```
# ENCODER (two hidden layers)
Linear(in_dim, hidden_dim)
LayerNorm()
ReLU()
Linear(hidden_dim, hidden_dim)
# Missing LayerNorm()
ReLU()

# ACTION HEAD (one hidden layer, followed by output layer)
Linear(hidden_dim, hidden_dim)
LayerNorm()
ReLU()
Linear(hidden_dim, out_dim)
```

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Thanks @MatthewCWeston for spotting that, @HassamSheikh could you fix that and add tests to ensure layernorm is included as expected

**Comment by HassamSheikh** [comment]
@MatthewCWeston I have updated the fix, please let me know if the current solution solves your problem

**Comment by MatthewCWeston** [comment]
@HassamSheikh Seems like it should solve the problem well; thanks

File: rllib/core/models/tests/test_encoder_layernorm.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+def get_layernorm_positions(sequential_net):
+    """Get indices and normalized shapes of LayerNorm layers in a Sequential net."""
+    result = []
+    for i, layer in enumerate(sequential_net):
+        if isinstance(layer, nn.LayerNorm):
+            result.append((i, tuple(layer.normalized_shape)))
+    return result
+
+
+class TestEncoderLayerNormConfig(unittest.TestCase):
```
Comment (pseudo-rnd-thoughts) [inline rllib/core/models/tests/test_encoder_layernorm.py]: I believe we agreed as a team to use pytest going forward. Could you modify all the tests to use pytest rather than unittest

File: rllib/core/models/tests/test_encoder_layernorm.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+    """Tests for LayerNorm in MLPEncoderConfig with various config combinations."""
+
+    def test_hidden_only_layernorm(self):
```
Comment (pseudo-rnd-thoughts) [inline rllib/core/models/tests/test_encoder_layernorm.py]: All of these tests look like the same test with different configs. If correct, update to use `pytest.mark.parametrize` to simplify the testing

File: rllib/core/models/tests/test_mlp_encoders.py [hunk 3]
```diff
@@ -18,6 +18,7 @@ def test_mlp_encoders(self):
         list_of_hidden_layer_dims = [[], [1], [64, 64], [256, 256, 256]]
         hidden_layer_activations = [None, "linear", "relu", "tanh", "swish"]
         hidden_layer_use_layernorms = [False, True]
+        output_layer_use_layernorms = [False, True]
```
Comment (pseudo-rnd-thoughts) [inline rllib/core/models/tests/test_mlp_encoders.py:21]: What is the difference between the hidden and output layer? 
In my head, you would never use layernorm on the final (output as I understand it) layer of a network.

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
As you added a test file, then you need to add it to the bazel. 

```
Exception: Cannot find bazel targets for tests ['python/ray/rllib/core/models/tests/test_encoder_layernorm.py'] so they won't be run automatically by CI, please add them to BUILD files.
```

Where is the layer norm applied in the MLP? Are we doing pre-norm or post-norm? 
`Input → LayerNorm → Linear (weights × input + bias) → Activation → Output`
or 
`Input → Linear (weights × input + bias) → Activation → LayerNorm → Output`

File: rllib/core/models/tests/test_mlp_encoders.py (hunk 3 above)
Comment (MatthewCWeston) [reply rllib/core/models/tests/test_mlp_encoders.py:21]: @pseudo-rnd-thoughts The encoder and the heads are instantiated separately - the change @HassamSheikh made prevents the last layer of the encoder from being incorrectly treated as an 'output' layer when it isn't one.

To my understanding, it does the same thing as my fix, it's just implemented with a new setting rather than a modification of the MLP construction logic.

**Review by pseudo-rnd-thoughts — APPROVED** [review]
LGTM, thanks for the improvements Hassam
